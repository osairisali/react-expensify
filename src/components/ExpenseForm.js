import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";
import { history } from "../routers/AppRouter";

export default class ExpenseForm extends React.Component {
  // pakai constructor agar bisa akses props yg diteruskan pada komponen ini dari EditExpensePage dan AddExpensePage
  constructor(props) {
    super(props);
    this.state = {
      description: props.expense ? props.expense.description : "",
      amount: props.expense ? (props.expense.amount / 100).toString() : "",
      note: props.expense ? props.expense.note : "",
      createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
      calendarFocused: false,
      error: "",
    };
  }

  // hanya menyimpan state sebagai object dlm class ExpenseForm, bukan pada constructor
  // state = {
  //   description: "",
  //   amount: "",
  //   note: "",
  //   createdAt: moment(),
  //   calendarFocused: false,
  //   error: "",
  // };

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({ description }));
  };

  onAmountChange = (e) => {
    const amount = e.target.value;
    // digit maksimal 2 angka sekaligus memastikan data yg diinput adalah number
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({ amount }));
    }
  };

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({ note }));

    // alternatif lain jika tanpa membuat variable note terlebih dahulu perlu pakai e.persist()
    // jk tdk, maka akan muncul error
    // e.persist();
    // this.setState(() => ({ note: e.target.value }));
  };

  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState(() => ({ createdAt }));
    }
  };

  onFocusChange = ({ focused }) => {
    this.setState(() => ({ calendarFocused: focused }));
  };

  onSubmit = (e) => {
    e.preventDefault();
    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: "Please provide description and amount!",
      }));
    } else {
      this.setState(() => ({ error: "" }));
      // return object data to props
      this.props.onSubmit({
        description: this.state.description,
        amount: parseFloat(this.state.amount, 10) * 100,
        note: this.state.note,
        createdAt: this.state.createdAt.valueOf(),
      });
    }
  };

  render() {
    return (
      <form className="form" onSubmit={this.onSubmit}>
        <p className="form__error">{this.state.error}</p>
        <input
          type="text"
          className="text-input"
          placeholder="Description"
          autoFocus
          value={this.state.description}
          onChange={this.onDescriptionChange}
        />
        {/* pakai prop step="any" agar bisa input floating number */}
        <input
          className="text-input"
          type="number"
          step="any"
          placeholder="Amount"
          value={this.state.amount}
          onChange={this.onAmountChange}
        />
        <SingleDatePicker
          date={this.state.createdAt}
          focused={this.state.calendarFocused}
          onDateChange={this.onDateChange}
          onFocusChange={this.onFocusChange}
          numberOfMonths={1}
          isOutsideRange={() => false}
        />
        <textarea
          placeholder="Add note for your expense (optional)"
          className="textarea"
          value={this.state.note}
          onChange={this.onNoteChange}
        ></textarea>
        <div>
          {/* {console.log(history)} */}
          {history.location.pathname.includes("edit") ? (
            <button className="button">Save Expense</button>
          ) : (
            <button className="button">Add Expense</button>
          )}
        </div>
      </form>
    );
  }
}
