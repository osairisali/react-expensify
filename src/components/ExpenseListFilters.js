import "react-dates/initialize";
import React from "react";
import { connect } from "react-redux";
import {
  setTextFilter,
  sortByAmount,
  sortByDate,
  setStartDate,
  setEndDate,
} from "../actions/filters";
import { DateRangePicker } from "react-dates";

//export class untuk test
export class ExpenseListFilters extends React.Component {
  // nggak usah pake constructor krn nggak ada props yg dimasukkan ke Component ExpenseListFilters dari ExpenseList
  // constructor(props) {
  //   super(props);
  // }

  state = {
    calendarFocused: null,
  };

  onDatesChange = ({ startDate, endDate }) => {
    this.props.setStartDate(startDate);
    this.props.setEndDate(endDate);
  };

  onFocusChange = (calendarFocused) => {
    this.setState(() => ({ calendarFocused }));
  };

  setTextFilter = (e) => {
    // dispatch dapat diakses melalui props selama dihubungkan ke store lewat connect()
    this.props.setTextFilter(e.target.value);
    console.log(e.target.value);
  };

  sortBy = (e) => {
    if (e.target.value === "date") {
      this.props.sortByDate();
    } else if (e.target.value === "amount") {
      this.props.sortByAmount();
    }
  };

  render() {
    return (
      <div className="content-container">
        <div className="input-group">
          <div className="input-group__item">
            <input
              className="text-input"
              placeholder="Search expenses"
              type="text"
              defaultValue={this.props.filters.text}
              // regiter function ketika terjadi perubahan pada input field
              onChange={this.setTextFilter}
            />
          </div>
          <div className="input-group__item">
            <select
              className="select"
              value={this.props.filters.sortBy}
              onChange={this.sortBy}
            >
              <option value="date">Date</option>
              <option value="amount">Amount</option>
            </select>
          </div>
          <div className="input-group__item">
            <DateRangePicker
              startDate={this.props.filters.startDate}
              endDate={this.props.filters.endDate}
              onDatesChange={this.onDatesChange}
              focusedInput={this.state.calendarFocused}
              onFocusChange={this.onFocusChange}
              numberOfMonths={1}
              showClearDates={true}
              isOutsideRange={() => false}
              startDateId={"startDate"}
              endDateId={"endDate"}
            />
          </div>
        </div>
      </div>
    );
  }
}

// ExpenseListFilters akan diimport oleh ExpenseList

// const ExpenseListFilters = (props) => (
//   <div>
//     <input
//       type="text"
//       defaultValue={props.filters.text}
//       // regiter function ketika terjadi perubahan pada input field
//       onChange={(e) => {
//         // dispatch dapat diakses melalui props
//         props.dispatch(setTextFilter(e.target.value));
//         console.log(e.target.value);
//       }}
//     />
//     <select
//       value={props.filters.sortBy}
//       onChange={(e) => {
//         if (e.target.value === "date") {
//           props.dispatch(sortByDate());
//         } else if (e.target.value === "amount") {
//           props.dispatch(sortByAmount());
//         }
//       }}
//     >
//       <option value="date">Date</option>
//       <option value="amount">Amount</option>
//     </select>
//   </div>
// );

const mapStateToProps = (state) => {
  // ambil default state dr filters untuk diteruskan sebagai props pada ExpenseListFilters
  return {
    filters: state.filters,
  };
};

const mapDispatchToProps = (dispatch) => ({
  setStartDate: (startDate) => {
    dispatch(setStartDate(startDate));
  },
  setEndDate: (endDate) => {
    dispatch(setEndDate(endDate));
  },
  setTextFilter: (text) => {
    dispatch(setTextFilter(text));
  },
  sortByDate: () => {
    dispatch(sortByDate());
  },
  sortByAmount: () => {
    dispatch(sortByAmount());
  },
});

export default connect(mapStateToProps, mapDispatchToProps)(ExpenseListFilters);
