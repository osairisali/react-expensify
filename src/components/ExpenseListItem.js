import React from "react";
import { Link } from "react-router-dom";
import moment from "moment";
import numeral from "numeral";

// dispatch bisa diakses melalui props ketika connect() ada
const ExpenseListItem = ({ description, amount, createdAt, id }) => (
  // dimasukkan dalam Link, jd semua di dalamnya (termasuk div) jd clickable
  <Link className="list-item" to={`/edit/${id}`}>
    <div>
      <h3 className="list-item__title">{description}</h3>
      <span className="list-item__sub-title">{moment(createdAt).format("MMMM Do, YYYY")}</span>
    </div>
    <h3 className="list-item__data">{numeral(amount / 100).format("$0,0.00")}</h3>
  </Link>
);

// tidak perlu pake mapStateToProps fun krn ini tdk berhubungan dengan state
export default ExpenseListItem;
