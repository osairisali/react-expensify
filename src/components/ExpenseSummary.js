import React from "react";
import numeral from "numeral";
import getExpensesTotal from "../selectors/expenses-total";
import selectExpenses from "../selectors/expenses";
import { connect } from "react-redux";

export const ExpenseSummary = ({ expenseCount, expenseTotal }) => {
  const expenseWord = expenseCount > 1 ? "expenses" : "expense";
  const formattedExpensesTotal = numeral(expenseTotal / 100).format("$ 0,0.00");

  return (
    <div>
      <h1>
        Viewing {expenseCount} {expenseWord} totalling {formattedExpensesTotal}{" "}
      </h1>
    </div>
  );
};

const mapStateToProps = (state, props) => {
  const visibleExpenses = selectExpenses(state.expenses, state.filters);
  return {
    expenseCount: visibleExpenses.length,
    expenseTotal: getExpensesTotal(visibleExpenses),
  };
};

export default connect(mapStateToProps)(ExpenseSummary);
