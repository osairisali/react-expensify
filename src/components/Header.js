import React from "react";
import { NavLink } from "react-router-dom";
import { startLogout } from "../actions/auth";
import { connect } from "react-redux";

export const Header = ({ startLogout }) => (
  <header>
    <h1>Expensify</h1>
    {/* perhatikan di sini juga perlu pake exact:{true} jika hanya redirect ke "/" */}
    <NavLink to="/dashboard" activeClassName="is-active">
      Dashboard
    </NavLink>
    <NavLink to="/create" activeClassName="is-active">
      Create Page
    </NavLink>
    <button onClick={startLogout}>Logout</button>
  </header>
);

const mapDispatchToProps = (dispatch) => ({
  startLogout: () => dispatch(startLogout()),
});

export default connect(undefined, mapDispatchToProps)(Header);
