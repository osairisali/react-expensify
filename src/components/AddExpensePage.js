import React from "react";
import ExpenseForm from "./ExpenseForm";
import { connect } from "react-redux";
import { startAddExpense } from "../actions/expenses";

// dibuatkan class agar tidak render AddExpensePage berulangkali
// export AddExpensePage ini cuma untuk test case
export class AddExpensePage extends React.Component {
  // dibuatkan variabel berupa arrow fun sbg referensi eksekusi ketika addExpense prop
  // pada AddExpensePage dijalankan, pakai method biasa juga bisa tapi harus pake bind()
  // ini terkait referensi this pada method yg mengarah ke instance onSubmit ketika direferensikan
  // dengan this.addExpense tanpa binding pada <AddExpenseForm />
  startAddExpense = (expense) => {
    this.props.startAddExpense(expense);
    this.props.history.push("/");
    // console.log(Object.getOwnPropertyNames(this));
  };

  // sebagai method pada class pada umumnya
  // onSubmit(expense) {
  //   this.props.addExpense(expense);
  //   this.props.history.push("/");
  //   console.log(this);
  // }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm onSubmit={this.startAddExpense} />
        </div>
      </div>
    );
  }
}

// const AddExpensePage = (props) => (
//   <div>
//     <h1>Add Expense</h1>
//     <ExpenseForm
//       onSubmit={(expense) => {
//         console.log(expense);
//         props.dispatch(addExpense(expense));
//         // langsung redirect ke dashboard
//         props.history.push("/");
//       }}
//     />
//   </div>
// );

// mapDispatchToProps adalah arg kedua setelah arg mapStateToProps pada connect
// bekerja untuk return object addExpense ke props dari AddExpensePage
const mapDispatchToProps = (dispatch) => ({
  startAddExpense: (expense) => dispatch(startAddExpense(expense)),
});

// nggak usah pake mapStateToProps krn nggak ngambil state dari sini
// cuma nambahin state, jd nggak perlu ambil prevState dr store
export default connect(undefined, mapDispatchToProps)(AddExpensePage);
