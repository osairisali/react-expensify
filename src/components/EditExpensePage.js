import React from "react";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import { startEditExpense } from "../actions/expenses";
import { startRemoveExpense } from "../actions/expenses";

export class EditExpensePage extends React.Component {
  onSubmit = (expense) => {
    console.log("updated expense: ", expense);
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push("/");
    // console.log(this);
  };

  onRemove = (e) => {
    console.log("remove button clicked");
    // {id} dienkapsulasi dalam object krn parameter removeExpense butuh object yg didalamnya ada id
    this.props.startRemoveExpense(this.props.expense.id);
    this.props.history.push("/");
  };

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Edit expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm expense={this.props.expense} onSubmit={this.onSubmit} />
          <button className="button button__secondary" onClick={this.onRemove}>Remove Expense</button>
        </div>
      </div>
    );
  }
}

// const EditExpensePage = (props) => {
//   console.log(props);
//   return (
//     <div>
//       <ExpenseForm
//         expense={props.expense}
//         onSubmit={(expense) => {
//           console.log("updated expense: ", expense);
//           props.dispatch(editExpense(props.expense.id, expense));
//           props.history.push("/");
//         }}
//       />
//       <button
//         onClick={(e) => {
//           console.log("remove button clicked");
//           // {id} dienkapsulasi dalam object krn parameter removeExpense butuh object yg didalamnya ada id
//           props.dispatch(removeExpense({ id: props.expense.id }));
//           props.history.push("/");
//         }}
//       >
//         Remove
//       </button>
//     </div>
//   );
// };

// ada dua parameter yg bisa diambil: state dr store dan props dr parent Component Router (akses id dr path)
const mapStateToProps = (state, props) => {
  // console.log("state dari mapStateToProps: ", state, "props: ", props);
  // ini akan return object expense ke props pada EditExpensePage melalui connect()
  return {
    expense: state.expenses.find(
      // akses id untuk filtering expense dari path di Router lewat match.params.id
      (expense) => expense.id === props.match.params.id
    ),
  };
};

const mapDispatchToProps = (dispatch) => ({
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense)),
  startRemoveExpense: (id) => dispatch(startRemoveExpense(id)),
});

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);
