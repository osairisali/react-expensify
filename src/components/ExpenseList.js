import React from "react";
import { connect } from "react-redux";
import ExpenseListItem from "./ExpenseListItem";
import selectExpenses from "../selectors/expenses";

// Compnent ExpenseList akan diimpor melalui ExpenseDashboardPage pada AppRouter

// export const ExpenseList = (props) => (
//   <div>
//     <ExpenseListFilters />
//     {props.expenses.length === 0 ? (
//       <p>No Expenses</p>
//     ) : (
//       props.expenses.map((expense) => {
//         return <ExpenseListItem key={expense.id} {...expense} />;
//       })
//     )}
//     {/* yg bs dirender hanya jsx dan primitive value js, render object akan
//     return error  */}
//   </div>
// );

export const ExpenseList = ({ expenses }) => (
  <div className="content-container">
    <div className="list-header">
      <div className="show-on-mobile">Expenses</div>
      <div className="show-on-desktop">Expense</div>
      <div className="show-on-desktop">Amount</div>
    </div>
    <div className="list-body">
      {expenses.length === 0 ? (
        <div className="list-item list-item__message">
          <span>No Expenses</span>
        </div>
      ) : (
        expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
      )}
    </div>
  </div>
);

// ini fun yg return object untuk digunakan sbg props pada ExpenseList di atas
// fun ini akan menghubungkan store dr redux sbg props pada component ExpenseList
const mapStateToProps = (state) => {
  return {
    // bisa juga return key-value pair yg bukan dr state store
    // name: "ali",
    expenses: selectExpenses(state.expenses, state.filters),
  };
};

// setiap store berubah, maka komponen akan ikut dirender ulang
export default connect(mapStateToProps)(ExpenseList);
