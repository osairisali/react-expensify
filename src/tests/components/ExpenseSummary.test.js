import React from "react";
import numeral from "numeral";
import expenseData from "../fixtures/expenses";
import { ExpenseSummary } from "../../components/ExpenseSummary";
import { shallow } from "enzyme";

test("should render expense summary with 1 expense", () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={1} expenseTotal={235} />
  );
  expect(wrapper).toMatchSnapshot();
});

test("should render expense summary with multiple expenses", () => {
  const wrapper = shallow(
    <ExpenseSummary expenseCount={12} expenseTotal={20235} />
  );
  expect(wrapper).toMatchSnapshot();
});
