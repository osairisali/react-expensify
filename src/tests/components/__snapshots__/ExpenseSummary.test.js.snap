// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render expense summary with 1 expense 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        1
      </span>
       
      expense
       totalling
       
      <span>
        $ 2.35
      </span>
       
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Create Expense
      </Link>
    </div>
  </div>
</div>
`;

exports[`should render expense summary with multiple expenses 1`] = `
<div
  className="page-header"
>
  <div
    className="content-container"
  >
    <h1
      className="page-header__title"
    >
      Viewing 
      <span>
        12
      </span>
       
      expenses
       totalling
       
      <span>
        $ 202.35
      </span>
       
    </h1>
    <div
      className="page-header__actions"
    >
      <Link
        className="button"
        to="/create"
      >
        Create Expense
      </Link>
    </div>
  </div>
</div>
`;
